src/arithmetic/algebraic_expression.c

```diff
@@ -481,16 +466,82 @@ void AlgebraicExpression_Transpose(AlgebraicExpression *ae) {
-}
-
 void AlgebraicExpression_Free(AlgebraicExpression* ae) {
-    for(int i = 0; i < ae->operand_count; i++)
-        if(ae->operands[i].free)
+    for(int i = 0; i < ae->operand_count; i++) {
+        if(ae->operands[i].free) {
             GrB_Matrix_free(&ae->operands[i].operand);
+        }
+    }
 
     free(ae->operands);
     free(ae);
 }
+
+AlgebraicExpressionNode *AlgebraicExpressionNode_NewOperationNode(AL_EXP_OP op) {
+    AlgebraicExpressionNode *node = rm_malloc(sizeof(AlgebraicExpressionNode));
+    node->type = AL_OPERATION;
+    node->operation.op = op;
+    node->operation.l = NULL;
+    node->operation.r = NULL;
+    return node;
+}
+
+AlgebraicExpressionNode *AlgebraicExpressionNode_NewOperandNode(GrB_Matrix operand) {
+    AlgebraicExpressionNode *node = rm_malloc(sizeof(AlgebraicExpressionNode));
+    node->type = AL_OPERAND;
+    node->operand = operand;
+    return node;
+}
+
+void AlgebraicExpressionNode_AppendLeftChild(AlgebraicExpressionNode *root, AlgebraicExpressionNode *child) {
+    assert(root && root->type == AL_OPERATION && root->operation.l == NULL);
+    root->operation.l = child;
+}
+
+void AlgebraicExpressionNode_AppendRightChild(AlgebraicExpressionNode *root, AlgebraicExpressionNode *child) {
+    assert(root && root->type == AL_OPERATION && root->operation.r == NULL);
+    root->operation.r = child;
+}
+
+GrB_Matrix AlgebraicExpression_Eval(const AlgebraicExpressionNode *exp) {
+    if(exp == NULL) return NULL;
+    if(exp->type == AL_OPERAND) {
+        return exp->operand;
+    } else {
+        GrB_Matrix r = AlgebraicExpression_Eval(exp->operation.r);
+        GrB_Matrix l = AlgebraicExpression_Eval(exp->operation.l);
+
+        // Determine if left or right expression needs to be transposed.
+        GrB_Descriptor desc = NULL;        
+        AlgebraicExpressionNode *rightHand = exp->operation.r;
+        AlgebraicExpressionNode *leftHand = exp->operation.l;
+        
+        if(leftHand && leftHand->type == AL_OPERATION && leftHand->operation.op == AL_EXP_TRANSPOSE) {
+            if(!desc) GrB_Descriptor_new(&desc);
+            GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+        }
+        
+        if(rightHand && rightHand->type == AL_OPERATION && rightHand->operation.op == AL_EXP_TRANSPOSE) {
+            if(!desc) GrB_Descriptor_new(&desc);
+            GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+        }
+
+        // Perform operation, result overrides R.
+        switch(exp->operation.op) {
+            case AL_EXP_MUL:
+                assert(GrB_mxm(r, NULL, NULL, Rg_structured_bool, l, r, desc) == GrB_SUCCESS);
+                break;
+            case AL_EXP_ADD:
+                assert(GrB_eWiseAdd_Matrix_Semiring(r, NULL, NULL, Rg_structured_bool, l, r, desc) == GrB_SUCCESS);
+                break;
+            case AL_EXP_TRANSPOSE:
+                // Transpose is an unary operation which gets delayed.
+                assert( !(l && r) && (l || r) ); // Verify unary.
```

Comment:
These conditions can be replaced with `(l ^ r)`, though I understand if you prefer what's written for readability.

Reply:
Yeah I think the long conditional is a bit more readable